File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
     int mod = 1e9+7;
    
    long long int sum(TreeNode *root)
    {
        if(!root) return 0;
        return (root->val+sum(root->left)+sum(root->right));
    }
    
    long long int dfs(TreeNode *root,long long int &s, long long int &ans)
    {
        if(!root) return 0;
        
         long long int l = dfs(root->left,s,ans);
         long long int r = dfs(root->right,s,ans);
        
        ans = max(ans,((s-l)*l));
        ans = max(ans,((s-r)*r));
      
        
        return (l+r+root->val);
    }
    
    int maxProduct(TreeNode* root) 
    {
        long long int ans = 0;
        long long s = sum(root);
        
        dfs(root,s,ans);
        return ans%mod;
        
    }
};

File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

var sumt = function(node)
{
    if(!node) return 0;
    return node.val+sumt(node.left)+sumt(node.right);
}

var dfs= function(node,s,ans)
{
    if(!node) return 0;
    
    const l = dfs(node.left,s,ans);
    const r = dfs(node.right,s,ans);
    
    ans.res = Math.max(ans.res,l*(s-l));
    ans.res = Math.max(ans.res,r*(s-r));
    
    return(node.val+l+r);
}

var maxProduct = function(root) 
{
    let mod = Math.pow(10,9)+7;
    let s = sumt(root);
    console.log(s)
    let ans  = {"res":0};
    dfs(root,s,ans);
    return ans.res%mod;

};